src/Compiler/Microsoft.AspNetCore.Razor.Language/src/DefaultBoundAttributeParameterDescriptorBuilder.cs

```diff
@@ -1,65 +1,55 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-#nullable disable
-
-using System;
 using System.Collections.Generic;
-using System.Linq;
+using System.Collections.Immutable;
+using Microsoft.AspNetCore.Razor.PooledObjects;
 
 namespace Microsoft.AspNetCore.Razor.Language;
 
-internal class DefaultBoundAttributeParameterDescriptorBuilder : BoundAttributeParameterDescriptorBuilder
+internal class DefaultBoundAttributeParameterDescriptorBuilder : BoundAttributeParameterDescriptorBuilder, IBuilder<BoundAttributeParameterDescriptor>
 {
     private readonly DefaultBoundAttributeDescriptorBuilder _parent;
     private readonly string _kind;
-    private readonly Dictionary<string, string> _metadata;
+    private readonly ImmutableDictionary<string, string>.Builder _metadata;
```

Comment:
❓ What is the access model for this?

* If the value is never converted to immutable (via `ToImmutable`), `Dictionary<TKey, TValue>` would be better here
* If the value is primarily read after converted to immutable, or if the mutations that occur are either rare or involve changing a majority of items in the collection, `ImmutableSegmentedDictionary<TKey, TValue>.Builder` would be better here
* If the value is frequently mutated after converted to immutable and those mutations are small, this collection is the right one to use

Reply:
It's always converted via `ToImmutable()`, however it is often empty. Previously, a new empty `Dictionary<string, string>` was created for each one of these.

Note that we don't have access to `ImmutableSegmentedDictionary` in Razor.